### DISCUSSION TITLE
Porting UI to Qt / GTK3

### DISCUSSION THREAD

**Discussion by iorsh** [fontforge/fontforge#5324] [Ideas]
Context: https://github.com/fontforge/fontforge/discussions/5298
> there are grassroot efforts to modernize its UI by porting it to GTK3. The current implementation uses custom UI toolkit (25 years old) with GDK under the hood, and it's really difficult to maintain or enhance.
>
> The initial GTK3 effort is done by @Omnikron13 at https://github.com/Omnikron13/fontforge/tree/gtk3-hybrid and by me (@iorsh at https://github.com/iorsh/fontforge/tree/recent_files)

**Comment by iorsh** [comment]
> I recommend separating the question of GTK3 from the question of browser support, or at least putting the questions in that order. A GTK3 port could help with in-browser use but I doubt there would be support for just abandoning the desktop client, so the first stage would be just getting GTK3 working in the desktop context.
> 
> I haven't looked at the above links, but one question is how complete these ports are. FontForge has many, many interfaces, including a large number of modal dialogs. Were _all_ of those ported to GTK3, or just a subset of them that the developer thought were relevant? Does the bitmap font window work, along with its dialogs? If not, you would have to get buy-in on abandoning those facilities, and it's very unclear how one would get buy-in on that given the distributed and largely silent nature of the user community.
> 
> You might be better off attempting to fork and to gain separate support for your fork. However, be warned: there have been many past partial forks of FontForge that have failed.

@skef, I agree with your points. The links above are just some initial attempts to check the feasibility of the entire endeavor. @Omnikron13 introduced GTK File Chooser, and I expanded it to better match the functionality of the current GDraw chooser.

In a separate effort I'm trying to convert the Font View to GTK (see https://github.com/iorsh/fontforge/tree/font_view). It's a very raw draft, but I could figure out the custom drawing of the character grid into a GTK widget,which seems to be one of the tougher pieces.

My idea is to have all the main views ported (Font, Character, Bitmap, Metrics) together with File Chooser and maybe some more stuff. I don't think the complete overhaul is feasible at all, it's just a really huge pile of UI. The above list though can qualify IMO for a Beta status.

One of (my personal) guidelines is losing NO functionality in the process. Nothing pisses people off more than stuff given to them and then taken away.

**Reply by skef** [reply]
I understand the spirit, but something is needed to reconcile these last two paragraphs: a "complete overhaul" isn't feasible, but no functionality should be lost. What will that mean in practice? A remaining compatibility layer, much like what is there now, for most of the dialogs? Maybe that could work, but at a minimum you'll probably have to rewire everything in the UI to use GTK3 style events, which is still a large undertaking.

**Reply by iorsh** [reply]
I mean it would be some hybrid application, where the e.g. "Font Info" dialog could be shiny new, but when clicking "Add Lookup" button, the old GDraw subdialog would pop up.

Thankfully, a huge work was done on GDK which basically allows us to mix toolkits and events nearly seamlessly.

**Reply by Omnikron13** [reply]
To be fair, in a lot of cases functionality is likely to be _gained_ rather than lost in a fair few areas.

Some existing dialogue crash too often to be reliable, some technically should be rendering UI elements which they simply aren't, some functionality is awkwardly split across multiple dialogs and buried in unintuitive usage, etc.

For example, I have been prototyping combining the two 'Groups' dialogs into a unified interface (albeit in a separate repo not actually hooked into the actual program at all for now). The existing UI is prone to seg faults when encountering empty groups, it has no capacity to rearrange created groups, it gets itself into a 'race' between the user and an error popup on invalid input, and the code for actually keeping track of the list of groups is broken and loses track of itself if you delete a group rendering it dangerous to attempt to continue using until it reloads.

All of these issues, along with e.g. merging the create/edit dialog with that for actual filtering, separating out categorisation of glyphs by name or code point, improving the management and display of group contents, potentially managing (save/load) different group sets, allowing the dialog to maintain itself 'always on top', etc. are relatively easily addressed, imo massively improving that function of the program.

(Integrating into the labyrinth of undocumented structs and functions that store and work with the group data is somewhat more of a headache, of course, and something of another Titan that IMO ultimately needs slaying for Fontforge to become a truly maintainable piece of software.)

**Comment by iorsh** [comment]
<img src="https://github.com/fontforge/fontforge/assets/3698866/7e1826fe-73ff-457d-9285-4af32375cc2d" width=500>

As of <a href="https://github.com/iorsh/fontforge/tree/13758e5a79e96d24e1fc61c67f45aa6a5916682a">now</a>, the Font View is bug-ridden (for example, the pop-up menu works fine, but the top menubar is still not operational).

But the screenshot gives some impression of the overall direction.

**Reply by Omnikron13** [reply]
There's absolutely enormous scope for UI improvements if and when the main font view can be smoothly liberated from the old UI codebase, not to mention the stepping-off point this will provide for anyone wanting to try and contribute code, so this is very promising!

Haven't really had the time to do much tinkering or spelunking into the codebase lately unfortunately, but I hopefully will be able to have a good look at this progress after some January university work is out of the way.

**Comment by Omnikron13** [comment]
The last I've had chance to really do anything on has just been prototyping sort of show-piece UI upgrades for more minor dialogs:
![ff_groups_ui_new_2](https://github.com/fontforge/fontforge/assets/2133262/d09479f4-6129-48d9-b5da-9b22d76da8ec)

Early days merging the two groups dialogs, trying to present their information more cleanly, enable rearranging the groups with drag&drop, etc. etc.

I promise I will actually integrate it into the actual program as soon as I have time to bring it up to full feature parity alongside the added bells and whistles. =P

**Comment by iorsh** [comment]
It would be good to reach some sort of consensus on this approach before we invest more work. For me, the main advantage here is that things can be done one piece at a time, instead of the all-or-nothing attempts which proved to be too laborious more than once. But there are probably other considerations which I may be not aware of.

@skef, what do you think about this? Maybe we shall also seek some initial feedback from other contributors?

**Comment by Omnikron13** [comment]
> For me, the main advantage here is that things can be done one piece at a time, instead of the all-or-nothing attempts which proved to be too laborious more than once.

This has been my reasoning as well. Approaching it as something of a 'Ship of Theseus' means that solid, usable improvements (to both functionality and stability) can be made relatively quickly and painlessly, and even if that starts with nibbling at the edges with bringing improvements to minor dialogs it does actually result in small real benefits to user experience and even some small bonuses to the overall codebase maintainability as not only the old arcane UI code gets stripped out but some of the periphery code around them becomes a _lot_ more accessible to tidy up.

> It would be good to reach some sort of consensus on this approach before we invest more work.
> [...]
> But there are probably other considerations which I may be not aware of.

Attempting to identify which areas are the least entangled into the codebase as good targets for relatively drop-in replacement is unfortunately something that seems pretty high on the priority list. I've done bits of digging about, though not much documenting of what I've found ultimately as it gets something of a labyrinth of chasing function calls and large mysterious structs about.

Colour picking is certainly one worth untangled, as colours are currently handled rather... strangely, but i don't believe they are too widespread.

Breaking up and documenting some of the mammoth functions in general would make things a lot easier for others to contribute too, to be honest. (As an example, the whole process of the program starting up, and reading command line arguments, etc. is paradoxically both composed of enormous dense functions, and tiny utility functions of a couple of lines in completely different files that seem to be called from just one or two places in the whole codebase. Dropping in a more standard solution for the arguments, integrating the tiny functions, commenting and a little refactoring, etc. would leave a much more streamlined entry point for anyone daring to try to get their bearings on the codebase.)

Documenting/commenting the various data structures is unfortunately perhaps another area that would make actually working on the UI a hell of a lot easier and more palatable. Anyone attempting to make any real modifications to anything is very quickly going to find themselves bogged down trying to work out what data is where and what it is doing, which is just replicated wasted time for each person really at present. The trouble there is they are rather arcane, so it's a case of either somebody already properly familiar with them doing it, or a hell of a lot of meticulous work spelunking around mapping them out.

**Comment by skef** [comment]
There are a lot of dimensions to these questions.

The main medium-term UI/widget issue is the current dependence on a particular version of GDK, where newer versions of GDK work differently and the current one doesn't work with Wayland. @jtanx had much of a GTK port working but became frustrated with aspects of their model. I believe that is https://github.com/jtanx/fontforge/tree/gtk-works2 . He then switched to qt had felt he had better luck, but I don't think that work isn't finished either: https://github.com/jtanx/fontforge/tree/qt . Note that when I say "port" in this sense, I mean what @jtanx already did with GDK: Getting custom FontForge widgets to, in effect, work on top of some other "portable" graphics layer. 

Porting this or that window to "full GTK" without the others doesn't resolve that question.

As for an incremental approach, a few years ago I suggested adding some code to allow "full GTK" windows to coexist with gdraw windows so that new additions to the UI, which might include plugins, could just use GTK. I did a proof-of-concept in https://github.com/skef/fontforge/tree/gtk-bridge . That idea was rejected, and at least one reason for the rejection was that it would tie the project to GTK and there was some preference for leaving that question open. (@jtanx already had reservations at that point.)

Let's set that aside. In my personal view the main "customer-facing" issue raised by a partial GTK port is UI consistency. FontForge has an extensive UI customization system. GTK is also customizable, but in somewhat different ways and via different mechanisms. So any attempt at replacing, say, the font view with a GTK window raises the question of how the current customization system will interact with the window. E.g. will the overall look and functioning of the program be consistent between the two widget sets?

However, even if that's my "line" I'm not going to make a unilateral decision about such a large change, and the other central members of the project tend to be more conservative than I am. Porting one window to the GTK puts the project on a certain trajectory that limits other future options. Some of "us" may object just on that basis. 

Other stuff:

The discussion above mentions code stability. When it comes to the current system I'm mostly unmoved by those concerns. Yes, lots of FontForge code is bug-ridden. At this point the chance of hitting a crash is proportional to how much a given subsystem gets used. Most of the "mainstream" facilities work OK. I suppose the problem raised by the FontView is that it needs new functionality, including new widgets and existing widgets working in new ways, and that would be a pain to get working in gdraw. I don't disagree, but while that weighs on some of the things I've said above, it doesn't definitively change them. And, of course, [bugs can be fixed](https://github.com/fontforge/fontforge/pulls?page=1&q=is%3Apr+is%3Aclosed+author%3Askef).

Meta:

Note, as I have said in other contexts, that anyone who wants to can, of course, just fork the code and do exactly what they see fit without dealing with me or any other FontForge hoo-ha. That poses two challenges to the forker, one being the right kind of challenge and one being the wrong kind. The right kind of challenge is that to succeed the fork would have to be clearly better than the code in this repo, rather than just different in the way certain people felt it should be different, with everyone else dragged along by inertia. The wrong kind of challenge is the flip-side of that inertia: the fork could be better but maybe users wouldn't notice or wouldn't bother changing to it anyway. 

I say this to make it clear what this discussion is about, and will inevitably be about. It's not ultimately about code or direction, because you can write whatever code you want and take that code in whatever direction you want. The discussion is ultimately about a name and the inertia associated with that name. And the board of the foundation that has rights to the name, and thus this repository, has tended to be pretty conservative about the direction. In particular they tend to be much more focused on the needs and workflows of current users than on potential new users. That's how the cookie has crumbled and will probably how it will continue to crumble.

**Reply by skef** [reply]
My reference for the level of the conservatism of the project is #3638 . 

@davelab6 (who is himself a FontForge board member) had worked with others to come up with a much more rational menu layout. I saw a partial PR for the changes in 2019 and finished the implementation. It turned out that the idea had been rejected in an email (not noted in the PR) because some existing users might not like the change. It was felt that until the menu layout was made fully configurable, so that users could choose between the old and new options, it shouldn't be accepted. I was pissed! But that was life.

**Reply by iorsh** [reply]
At this particular moment, limiting the future of FontForge by tying it to some particular toolkit seems purely theoretical. Any future is better than zero commits in a year.

I agree that the hybrid approach effectively kills the consistency, theme customization and all the stuff around it. My opinion is that theme customization should have never been a concern at the first place. FontForge is a program for fonts, and it's functionality isn't affected by color of menu background. George Williams chose to write his own toolkit because he needed Unicode support, and back in 2000 no standard toolkit had any reasonable support for it.

On the other hand, the board has the last word, that's their right.

**Reply by Omnikron13** [reply]
To add my own comments to what @iorsh has said here;

FontForge is _already_ tied to a particular toolkit. The issue is that the toolkit it is tied to is bespoke and was written by, as far as I can tell, someone who does not now and has no future intention of ever again maintaining it. It is old. It is buggy (to the point it seg faults and crashes the whole program sometimes, doesn't display controls it is seemingly supposed to, etc.). I'm not sure there is anyone who knows how it works or what it is doing all the time. It is not commented or documented. It can only be used via a very low-level API, which adds to an already very verbose and Byzantine codebase.

IMHO any talk of consistency and theming is, frankly, laughable in the face of these more pressing concerns.

As the only serious contender as a font editor in the open source ecosystem, the current issues with the GUI code (as part of general issues with the aging of the codebase) are (or should be) a real concern to all who hold firm beliefs in the promise and importance of open source software.

The only options when it comes to making sure there continues to be a serious option for creating and editing fonts using open source software long into the future are to either create a new program from scratch written in modern and maintainable code (some have tried, and I'm sure you can find many aborted attempts around the internet), to fork it entirely independent of the main project (similar issues exist here; my own spare-time tinkerings which I can only very infrequently make time for time for are probably a good example of how that goes without a concerted effort), or to make serious attempts to modernise the existing codebase with a view to future maintainability.

The powers that be obviously have final say on what is or isn't merged into the mainline codebase, and it is quite right that this _excludes_ any attempts at large scale changes like this which do not manage to gain the momentum to reach completion. Adding inconsistency and unfinished changes to a venerable codebase like this is the last thing it needs.

On the flipside, a serious effort to replace the entire GUI codebase with something that new developers stand a chance of contributing to, maintained and worked on as an aside to the mainline codebase (ideally as a branch of the main repo rather than any other random developers fork, though that is a whole separate discussion), is something that FontForge is ultimately going to _need_ sooner or later.

**Comment by jtanx** [comment]
I haven't read the full thread but actually my qt branch more or less works. I'd definitely say that's the way forward if we ever need to move, especially as it has somewhat usable Wayland support

**Reply by iorsh** [reply]
@jtanx, unfortunately I discovered your qt branch just a week ago, quite by chance. Doing more research on my part before jumping in could have been the better thing to do. Sorry about that.

On the choice between Qt and GTK:
- I'm not a big fan of Qt, as I happened to work with it for 3 years, and it's mostly usable with its own flavor of C++ with slots and signals and all that stuff. Committing to a particular non-standard compiler sounds to me way worse than committing to a particular UI toolkit.
- If GTK is good enough for Inkscape and GIMP, then it's good enough for FontForge.
- Qt is open-source, but still proprietary, so if its owners decide to close it down, there wouldn't be any community in place to keep FOSS development.

On the negative side
- My opinion is obviously skewed, as I'm already emotionally invested in my own code (surprise, surprise...)
- I read a few recent rants about GTK4 that it's developers take a strong pro-mobile stance with hamburger menus and all that touchscreen stuff, effectively neglecting the desktop applications.

@jtanx, a question for you: I understand that your work in https://github.com/jtanx/fontforge/tree/gtk-works2 was abandoned due to some technical issues with GTK/GDK. I compiled it, and it actually looks good and working. What were the specific problems which held you back? Maybe there is a chance they have been resolved by now, or could be overcame?

**Reply by jtanx** [reply]
What's your main platform, Linux? Have you ever targeted Windows or Mac? GTK support is pretty trash, only made worse from GTK3 to GTK4. On the other hand, Qt acutally works, _and_ it even looks native, unlike GTK which either seems stuck in 1997 or have abandoned even trying to look native at all.

If you've looked at my branch, you'll actually notice I don't use slots/signals to port over the existing functionality. Rewriting new dialogs in Qt, then yeah probably, but I don't really see that as being too bad. 

> Committing to a particular non-standard compiler sounds to me way worse than committing to a particular UI toolkit.

I'm sorry, that's just misleading. Fundamentally the use of slots/signals implies you're writing Qt relevant code which you wouldn't be otherwise, so this is in no way worse than any other UI toolkit. Even then, we're talking about running code through automoc or whatever it's called, _not_ tying ourselves to a particular version of clang/gcc or whatever.


> Qt is open-source, but still proprietary, so if its owners decide to close it down, there wouldn't be any community in place to keep FOSS development.

Yeah, no. There is a GPL licensed version of it, if they decided to make it fully closed source going forward, I have no doubt someone would decide to fork it.

> If GTK is good enough for Inkscape and GIMP, then it's good enough for FontForge.

I could point you to counter-examples, like Wireshark/mkvtoolnix switching to Qt from GTK/WxWidgets respectively.

>  I understand that your work in https://github.com/jtanx/fontforge/tree/gtk-works2 was abandoned due to some technical issues with GTK/GDK. I compiled it, and it actually looks good and working. What were the specific problems which held you back? Maybe there is a chance they have been resolved by now, or could be overcame?

Have you actually tried running this under wayland?

**Reply by iorsh** [reply]
Ok, I see. In the defense of GTK3 looks, it supports custom themes, so giving it a native look is probably a matter of theme in use. I don't have Windows machine to test, but on Linux themes do work and give it some Win-like appearance.

On this occasion, I rescind my general resistance to UI customization. I still think it's not a business of any particular application, but supporting the existing FontForge customization through CSS in GTK or QSS in Qt now looks to me achievable with a reasonable effort.

Regarding Wayland, I understand that a lot of software uses GTK and XWayland, and it wouldn't be a really terrible bet just to let others blaze the trail.

To the point, with all the philosophy above aside. @jtanx, if your choice is Qt, what is the state of https://github.com/jtanx/fontforge/tree/qt then? Is there some list of TODOs which needs to be checked before the work is considered complete? Anything I can help with?

**Reply by jtanx** [reply]
Last I recalled, it basically worked, although it'd need wider usage to actually iron out remaining bugs. Given it's an option, I guess you could just merge it, same as what happened last time for the GDK support

I'd probably just drop Qt5 support for only Qt6. The other thing that I didn't get around to is how you'd actually integrate pure-Qt windows/dialogs while also running the GDraw based dialogs (i.e. how you'd actually replace the latter with the former).

I also got the IME working on Windows but I've never been able to get a working IME setup on Mac/Linux, so maybe also something worth looking into

There was https://github.com/users/jtanx/projects/1 but there's not much on it

I just don't really have time at the moment to properly work on fontforge, so best I can offer right now is just advice

**Reply by iorsh** [reply]
I checked the Qt port, my notes are as follows:

1. Seems to work overall. I didn't test it extensively, noticed a bug which causes all drawings to appear slightly scaled:
![image](https://github.com/fontforge/fontforge/assets/3698866/4df5c052-8526-4bbb-bf65-976686ea1105)
Looks like the cairo surface is scaled for some reason.

2. IME works (Kubuntu 22.04 + IBus):
![image](https://github.com/fontforge/fontforge/assets/3698866/fc0a2dde-07ea-4993-bfb0-80ee44445bbb)

3. I'll need to check combination with a pure-Qt dialog. @skef, do you have a candidate which is simple and sufficiently used, but wouldn't stand out too much when converted? Maybe validation results window?

**Reply by Omnikron13** [reply]
> The other thing that I didn't get around to is how you'd actually integrate pure-Qt windows/dialogs while also running the GDraw based dialogs (i.e. how you'd actually replace the latter with the former).

This would seem something of a sticking point, if I'm understanding you correctly?

Are you saying it would be much of an all-windows-or-nothing approach?

I'm certainly not familiar with Qt, so, is this likely to be a particularly difficult problem to solve?

**Reply by skef** [reply]
@Omnikron13 Based on past conversations I don't think @jtanx intended to say anything either way about how hard it would be, just that there's no code in the branch dealing with that.

With either a GTK or a Qt port one of the main challenges is keeping the UI at least somewhat uniform, which in practice either means replacing the current UI customization stuff with a more Qt-oriented layer or getting Qt to look more or less like the customized FontForge windows. People do care about that customization stuff so there would be resistance to removing it without some analogous replacement.

**Reply by Omnikron13** [reply]
> I checked the Qt port, my notes are as follows:

Any advice on getting it to build? Is there a particular commit I should be building from, or some settings I need to tweak first?

I had no luck at all which a quick attempt the other day, though I'll have to try again to see what errors I was getting.

A quick step-by-step of how you got things running would be appreciated though, to make sure I'm even trying the same thing.

**Reply by Omnikron13** [reply]
> @Omnikron13 Based on past conversations I don't think @jtanx intended to say anything either way about how hard it would be, just that there's no code in the branch dealing with that.

Ah I see, no worries then. Just misreading on my end by the sound of it.

> With either a GTK or a Qt port one of the main challenges is keeping the UI at least somewhat uniform, which in practice either means replacing the current UI customization stuff with a more Qt-oriented layer or getting Qt to look more or less like the customized FontForge windows. People do care about that customization stuff so there would be resistance to removing it without some analogous replacement.

There's a fair talk about getting things to look/feel/behave how people like, customisation wise. Perhaps it's worth trying to assemble an at least semi-comprehensive list of the specific things that would be real sticking points if lacking in any future customisation options?
(I personally would have thought there would be more options for customisation and such options, with a toolkit that more people are realistically able to work on, but, that's all speculation on my part really without some more concrete examples of the 'must have' options.)
Where exactly to assemble such a list, and how to get a good cross section of feedback from users I'm not so such though...

**Reply by iorsh** [reply]
> Any advice on getting it to build? Is there a particular commit I should be building from, or some settings I need to tweak first?

The following worked for me (already in `build`, reconstructing from history, hopefully works) :
```
git remote add jtanx https://github.com/jtanx/fontforge
git fetch https://github.com/jtanx/fontforge
git checkout qt
sudo apt install qt6-base-dev
cmake -GNinja .. ; ninja
```

**Reply by Omnikron13** [reply]
> The following worked for me (already in `build`, reconstructing from history, hopefully works) :
> ```
> git remote add jtanx https://github.com/jtanx/fontforge
> git fetch https://github.com/jtanx/fontforge
> git checkout qt
> sudo apt install qt6-base-dev
> cmake -GNinja .. ; ninja
> ```

Cheers, I'll try doing exactly that when I'm next at my laptop.

**Reply by Omnikron13** [reply]
My attempts to build on an arch-based box, after just temporarily installing the full `qt6` metapackage (absolutely no idea what would or wouldn't be required specifically) result in a series of errors of the form:
`/[..]/fontforge/po/fr.po:294: 'msgstr' is not a valid C format string, unlike 'msgid'. Reason: In the directive number 3, the argument size specifier is invalid.`
...which ultimately ends in a failed build. This seems to be translations or somesuch, however, so perhaps there is a more minimal build I should attempt?

**Reply by skef** [reply]
You'll either need to rebase or copy the po files from a more recent commit. This problem was fixed in https://github.com/fontforge/fontforge/pull/5330 and has to do with the pickiness of more recent toolchains.

**Reply by jtanx** [reply]
I've rebased the qt branch to master fwiw

**Reply by Omnikron13** [reply]
No issues at all building it on my box now.

**Comment by skef** [comment]
> do you have a candidate which is simple and sufficiently used, but wouldn't stand out too much when converted? Maybe validation results window?

Probably "Add Encoding Slots" in the "Encoding" menu. Frequently used but mostly by people who have long ago stopped thinking about it. Also very simple.

**Comment by skef** [comment]
So, just to clarify how this would probably work on a project level if it were to go forward:

Assuming any remaining Qt issues (like the scaling above) are addressed, and the powers that be are amenable (which seems likely -- @jtanx's views on gdraw are generally trusted), there would probably need to be a release that only makes the Qt changes to gdraw (plus whatever other non-UI stuff has been merged), and that will need to be out in the world for quite a while before any Qt-only UI stuff is accepted into the master branch. That way we could be more assured that this path for the UI code, or if it won't we can back out to GDK for the time being.

**Reply by iorsh** [reply]
This sounds reasonable to me, with a couple of comments.
1. It would be better to have this plan explicitly pre-vetted with the _powers that be_. The general idea and the technical direction seem to be clear as of now, and it would be counterproductive to proceed with bugfixes etc, if there is no consensus on the project substance.
2. After the initial release there should an agreed bug-free window of predetermined duration (with bux-fix releases if necessary), after which further transition to pure Qt will be acceptable.
3. I prefer to put the pure-Qt window in the release, active by default, with a preference (maybe config-only, without UI) to disable it in favor of the GDraw-based one. This would allow wider testing of pure-Qt path too, with a contingency plan in case of some major problem with that specific window. Does this sound ok?

**Reply by skef** [reply]
> It would be better to have this plan explicitly pre-vetted with the powers that be.

I don't disagree with this but that doesn't mean it will happen. It hasn't tended to happen in the past. There are some valid reasons for this and perhaps some not-so-great reasons.

**Comment by JoesCat** [comment]
Gonna throw in food-for-thought here:
[QT6 Supported Platforms](https://doc.qt.io/qt-6/supported-platforms.html)
[GTK4 Supported Platforms](https://www.gtk.org/docs/installations/index) [Debian as a reference](https://packages.debian.org/search?keywords=libgtk-4-1)

I do like KDE (3,4,5,6) as a Desktop, but it looks like it's narrowing-down to just the most popular desktop platforms, while GTK seems to still try and target what it can.

**Reply by skef** [reply]
Seems like this might be apples and oranges. In open source land "supported" means something like "will (probably) run". In the commercial software world it means "can pay the company to fix problems and get help". That first page above says "Supported platforms are actively maintained, subjected to testing, and bugs discovered in these platforms are prioritized for correction. Significant errors discovered in testing can impact release dates.", implying that there are other platforms it runs on.

**Reply by JoesCat** [reply]
Found some other info....
[QT6.6](https://tracker.debian.org/pkg/qt6-base) = [poppler](https://tracker.debian.org/pkg/poppler) and [qca2](https://tracker.debian.org/pkg/qca2) appears to be having some relatively good success. That's a good sign.

**Comment by davelab6** [comment]
I just wanted to drop by to say, if anyone cares what I think, that I think this is a very good idea and I would like to see it happen, and support anything that helps expedite it.

**Reply by davelab6** [reply]
I was reminded of this thread while just now reading https://betanews.com/2024/01/16/betterbird-the-email-client-thunderbird-could-have-been/ - the bit about profiles is intriguing. Perhaps fontforge with gdraw could be released as fontforge classic, and the qt version could be set up to run alongside it. That might free up the new effort so users who depend on classic features always have them

**Reply by iorsh** [reply]
Adopting a new UI toolkit is a major change (perhaps one of the biggest ever), and I believe it's worth planning, followed by clear decisions by the people who have the authority to take such decisions. As part of this planning, the stakeholders can review the features which have the biggest chance to get broken by the change, consider the risks and the mitigation actions. Such actions can range between dropping the feature and keeping it and prototyping it in the new environment to reduce risks, as deemed appropriate.

Of course, anything on the scale between deep review and no review can be done, as long as decisions are made.

**Comment by matj1** [comment]
GTK4 has been out for some time. Why not port FontForge interface to GTK4 instead of GTK3? I don't have experience with GTK, but it seems that GTK4 would be more future-proof and have more features because it is newer.

**Reply by iorsh** [reply]
Honestly, I just took whatever @Omnikron13 started and continued from there. From my limited experience with documentation, GTK4 is richer, but sometimes more complicated. Besides that, there seem to be more code examples for GTK3, and it helps when copy-pasting stuff.

**Reply by jtanx** [reply]
I already mentioned it above, gtk has pretty terrible cross platform support and it's only gotten worse in gtk4. There have been numerous times I have had to investigate quite deeply into the gdk/gtk source itself to figure out why it's been (mis)behaving.

They also deprecated a bunch of APIs with seemingly no viable replacement. It works ok enough as long as you stick to Linux and the gtk ecosystem, but as soon as you step outside of that you're in for a world of hurt

**Reply by Omnikron13** [reply]
>  Why not port FontForge interface to GTK4 instead of GTK3? I don't have experience with GTK.

It has been some time since I first tinkered about with the notion, but in short, as I recall, because the Gdk version that FF will build on is quite specific and old and there's no chance of getting GTK4 to play nice with it.

Now, the alternative could be to dig into the guts of the UI and port it to some suitable version of Gdk. This, imo, would be quite a pain. Iirc there's quite a few bits of Gdraw you'd have to muck about with.

The problem _there_ is that the codebase for Gdraw is not really commented  or documented, and you will likely spend quite some time bouncing from file to data structure to file to data structure (none of which likely to be commented or formatted). The codebase and documentation for Gdk is marginally better, but in all honestly largely expected to be used via GTK, and so working out what you need to replace with what to port a (20 year old?) 3rd party minimalist UI toolkit (whose original author I think stopped working on the project some 10 years ago?) doesn't strike me a a very fun or productive pet project.

Possible, I'd expect? Worthwhile? It seems doubtful to me.

GTK3 however is basically already pulled in with the version of Gdk that FF can build on. A little, admittedly hacky, tinkering and you can have GTK3 components playing _reasonably_ nicely alongside Gdraw ones, sharing an underlying Glib event loop.

Horribly broken and clunky to use UI dialogs can be, with a little more hacking about, replaced with GTK3 ones which are less buggy, much less likely to randomly segfault, provide better and less awkward (working) functionality, and be infinitely more maintainable given the number of people familiar with GTK Vs Gdraw.

> it seems that GTK4 would be more future-proof and have more features because it is newer.

The lifecycle of GTK is, to put it bluntly, kinda screwy. The 'latest' major version is, as I'm lead to believe, considering by the Devs themselves to be considered unstable for some considerable time (I forget what versions they actually recommend it be considered stable). The upshot is that hanging back a while is generally for the best outside new projects. With talk of work on GTK5, I'd say we're past that sticking point tbh.

_However_, as I say it is likely much easier to replace components piece by piece while still having a fully functional piece of software, even if it is something of a ship of Theseus.

If and when the UI were to be fully onboard with GTK3, then porting up to GTK4 would likely be _much_ less of a headache.

**Comment by iorsh** [comment]
I made some proof-of-concept with Font View screen fully rewritten in GTK3. As mentioned earlier, it's based on @Omnikron13's major infrastructural breakthrough.

You can check the code at https://github.com/iorsh/fontforge/tree/font_view or pick Linux AppImage at https://github.com/iorsh/fontforge/actions/runs/10101284086 (scroll to artifacts at the end of the page).

The code is very unsightly, I don't think I bothered to release a single allocated byte of memory there. So the next step is probably cleaning and brushing it.

It was tested in Linux (both X11 and Wayland); I still have to look how it behaves on Windows. @jtanx promised a lot of pain there, and I fully trust his experience.

There are notes in the [Wiki](https://github.com/iorsh/fontforge/wiki/TODOs)

**Reply by iorsh** [reply]
To those who tried to run this on Windows and unsurprisingly failed, please pick the fixed version at https://github.com/iorsh/fontforge/actions/runs/10801650837 (scroll to artifacts, I tested MINGW64). The branch should also compile fine in MSYS2.

**Comment by TNWC14** [comment]
Terribly sorry for reviving this but I don't want to start a new thread and I'd like to know if there is still work being done on this or maybe in the plans at least?

**Comment by iorsh** [comment]
Yes, it was decided to proceed with porting UI to GTK3, and the latest release includes one GTK-based dialog for "Encoding --> Add Encoding Slots...". There is also a "Language List" dialog in #5648, and two pending PRs #5682, #5708.

The pace is not quite fast, as you can see, but we do advance in the right direction.

**Reply by TNWC14** [reply]
Oh okay, thank you for your work, it means a lot, it's good to see that FF keeps up thanks to yours' and other maintainers' help.

**Reply by jtanx** [reply]
Personally I still think going down gtk3 is a dead end/waste of time, but I'm also not working on this ATM so 🤷

**Reply by TNWC14** [reply]
Cause of the ancient codebase or just it's not worth the effort?

**Reply by iorsh** [reply]
I'm doing my best to localize the GTK UI code as much as possible, so that if the impending doom arrives, the catastrophe would be less catastrophic. Meanwhile @skef made the code much more C++-friendly in #5750, and a lot of my effort goes into disentangling UI toolkit in a neutral manner.

**Reply by TNWC14** [reply]
Wow, that's a lot of work done, I am quite speechless, thank you.

**Reply by skef** [reply]
It was mostly Claude, when it comes down to it.

**Comment by baturax** [comment]
hi, how far has it been doen

**Reply by iorsh** [reply]
A few dialogs have been ported to GTK3. I wouldn't say it's advancing fast, but it's definitely not stale. Any help would be appreciated!
